### PR TITLE
feature [nativelib]: implement `MemoryMXBean` and `ManagementFactory`

### PR DESCRIPTION
**Pull request by iRevive**
The first step towards https://github.com/scala-native/scala-native/issues/4018.

### Scope

1) Implement `java.lang.management.MemoryMXBean`
2) Implement `java.lang.management.MemoryUsage`
3) Implement `java.lang.management.ManagementFactory.getMemoryMXBean()`

### CODE REVIEW

File: javalib/src/main/scala/java/lang/management/ManagementFactory.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+   *  @example
+   *    {{{
+   *  val memoryBean = getMemoryMXBean()
+   *  println(s"current heap: ${memoryBean.getHeapMemoryUsage().getUsed()}")
+   *
+   *  val list = List.fill(Short.MaxValue)(0) // allocate memory
+   *  println(s"current heap: ${memoryBean.getHeapMemoryUsage().getUsed()}")
+   *    }}}
+   */
+  def getMemoryMXBean(): MemoryMXBean = MemoryBean
```
Comment (iRevive) [inline javalib/src/main/scala/java/lang/management/ManagementFactory.scala:18]: OpenJDK implementation - [ManagementFactory.java](https://github.com/openjdk/jdk/blob/master/src/java.management/share/classes/java/lang/management/ManagementFactory.java#L345-L347).

File: javalib/src/main/scala/java/lang/management/MemoryMXBean.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package java.lang.management
+
+import scala.scalanative.runtime.GC
+
+trait MemoryMXBean {
```
Comment (iRevive) [inline javalib/src/main/scala/java/lang/management/MemoryMXBean.scala:5]: OpenJDK implementation - [MemoryMXBean.java](https://github.com/openjdk/jdk/blob/master/src/java.management/share/classes/java/lang/management/MemoryMXBean.java).

A few differences: 
1) The trait doesn't extend `PlatformManagedObject` 
2) The following methods aren't implemented: `getObjectPendingFinalizationCount`, `getNonHeapMemoryUsage`, `isVerbose`, `setVerbose`

File: javalib/src/main/scala/java/lang/management/MemoryMXBean.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  private[management] def apply(): MemoryMXBean =
+    new Impl
+
+  private class Impl extends MemoryMXBean {
+    def getHeapMemoryUsage(): MemoryUsage =
+      new MemoryUsage(
+        GC.getInitHeapSize().toLong,
+        GC.getUsedHeapSize().toLong,
+        -1L,
```
Comment (iRevive) [inline javalib/src/main/scala/java/lang/management/MemoryMXBean.scala:34]: I didn't find a simple way to get (calculate) the committed memory across all garbage collectors.

File: nativelib/src/main/resources/scala-native/gc/none/gc.c [hunk 4]
```diff
@@ -52,6 +52,8 @@ size_t scalanative_GC_get_max_heapsize() {
     return Parse_Env_Or_Default("GC_MAXIMUM_HEAP_SIZE", getMemorySize());
 }
 
+size_t scalanative_GC_get_used_heapsize() { return -1L; }
```
Comment (iRevive) [inline nativelib/src/main/resources/scala-native/gc/none/gc.c:55]: Is there a way to get the total allocated memory? 
I guess we can add a static variable and explicitly increment it on allocations. But is there a better way perhaps?

File: nativelib/src/main/resources/scala-native/gc/boehm/gc.c [hunk 5]
```diff
@@ -61,6 +61,16 @@ size_t scalanative_GC_get_max_heapsize() {
...
 }
 
+size_t scalanative_GC_get_used_heapsize() {
+    struct GC_prof_stats_s *stats =
+        (struct GC_prof_stats_s *)malloc(sizeof(struct GC_prof_stats_s));
+    GC_get_prof_stats(stats, sizeof(struct GC_prof_stats_s));
+    size_t heap_sz = stats->heapsize_full;
+    size_t unmapped_bytes = stats->unmapped_bytes;
+    free(stats);
+    return heap_sz - unmapped_bytes;
```
Comment (iRevive) [inline nativelib/src/main/resources/scala-native/gc/boehm/gc.c:71]: Hm. Could we use `GC_get_heap_size()` instead?

File: javalib/src/main/scala/java/lang/management/MemoryMXBean.scala (hunk 2 above)
Comment (keynmol) [reply javalib/src/main/scala/java/lang/management/MemoryMXBean.scala:5]: @iRevive open jdk is GPL licensed, you cannot use its sources as a reference.

Comment (ekrich) [reply javalib/src/main/scala/java/lang/management/MemoryMXBean.scala:5]: Yes, just use the Javadoc as a specification. You can also write tests that can run against the JVM and Native just to make sure the values make sense in general.

File: javalib/src/main/scala/java/lang/management/MemoryMXBean.scala (hunk 3 above)
Comment (WojciechMazur) [reply javalib/src/main/scala/java/lang/management/MemoryMXBean.scala:34]: This one might be more difficult to track. On Windows we do manually commit memory provided by VirtualAlloc, but in the case of Unix we typically only reserve a maximal chunk of memory using mmap and committing is done automatically by the system on the first usage of every page. Access to this memory is only limited by Immix/Commix Heap blocks limit. 
I guess the best choice would be to fill it with usedHeap which is still correct regarding Javadocs: `committed will always be greater than or equal to used`

File: nativelib/src/main/resources/scala-native/gc/boehm/gc.c (hunk 5 above)
Comment (WojciechMazur) [reply nativelib/src/main/resources/scala-native/gc/boehm/gc.c:71]: Probably it's thread-safe alternative might be better to use https://github.com/ivmai/bdwgc/blob/c3c44d85081044707eb8fa4f0db39e1fe6834853/include/gc/gc.h#L841-L850 
However, current solutions seems to be fine.

Comment (WojciechMazur) [inline nativelib/src/main/resources/scala-native/gc/boehm/gc.c:71]: You might consider allocating the struct on the stack
```suggestion
    struct GC_prof_stats_s stats = {};
    GC_get_prof_stats(&stats, sizeof(struct GC_prof_stats_s));
    size_t heap_sz = stats.heapsize_full;
    size_t unmapped_bytes = stats.unmapped_bytes;
    return heap_sz - unmapped_bytes;
```

File: nativelib/src/main/resources/scala-native/gc/none/gc.c (hunk 4 above)
Comment (WojciechMazur) [reply nativelib/src/main/resources/scala-native/gc/none/gc.c:55]: Allocated memory is not tracked in none GC (it works by always allocating until there is no memory left) Adding a static variable to track allocations should do the job

File: javalib/src/main/scala/java/lang/management/MemoryMXBean.scala (hunk 3 above)
Comment (iRevive) [reply javalib/src/main/scala/java/lang/management/MemoryMXBean.scala]: Thanks for the explanations! Let's fill committed memory with the amount of used heap.

File: nativelib/src/main/resources/scala-native/gc/none/gc.c (hunk 4 above)
Comment (iRevive) [reply nativelib/src/main/resources/scala-native/gc/none/gc.c]: I added the static `TOTAL_ALLOCATED` variable to track the allocated memory.

File: nativelib/src/main/resources/scala-native/gc/boehm/gc.c (hunk 5 above)
Comment (iRevive) [reply nativelib/src/main/resources/scala-native/gc/boehm/gc.c]: Done, thanks.

**Comment by iRevive** [comment]
Are tests flaky, or did I break something?

**Comment by WojciechMazur** [comment]
> Are tests flaky, or did I break something?

One of the recently added tests was probably flaky. I was hopefully fixed on main right already, but is not applicable here without the rebase.

**Review by WojciechMazur — APPROVED** [review]

**Pull request merged**